File: gobench/provision.go
package gobench

import (
	"context"
	"fmt"
)

func rangeKeys(prefix string, count int) []string {
	keys := make([]string, 0, count)
	for index := 0; index < count; index++ {
		keys = append(keys, fmt.Sprintf("%s/obj-%04d.bin", prefix, index))
	}
	return keys
}

func seedKeys(ctx context.Context, clients *Clients, bucket string, keys []string, objectSizeBytes int, seedPrefix string) error {
	for _, key := range keys {
		body := CreatePayload(objectSizeBytes, fmt.Sprintf("%s:%s:%s", seedPrefix, bucket, key))
		if err := clients.PutObject(ctx, bucket, key, body); err != nil {
			return err
		}
	}
	return nil
}

func ProvisionRun(ctx context.Context, clients *Clients, options BenchmarkOptions) (BenchmarkManifest, error) {
	runID := MakeRunID(options.Prefix)
	artifactDir := ResolveArtifactDir(options.ArtifactsRoot, runID)
	if err := EnsureDir(artifactDir); err != nil {
		return BenchmarkManifest{}, err
	}

	mutableFixtureCount := options.Iterations + options.Warmup
	sharedReadKeys := rangeKeys("fixtures/shared/read", max(options.Iterations, options.SharedReadCount))
	localReadKeys := rangeKeys("fixtures/local/read", max(options.Iterations, options.LocalReadCount))
	overwriteKeys := rangeKeys("fixtures/shared/overwrite", max(mutableFixtureCount, options.OverwriteCount))
	deleteKeys := rangeKeys("fixtures/shared/delete", max(mutableFixtureCount, options.DeleteCount))
	listPrefix := "fixtures/shared/list/"
	listKeys := rangeKeys("fixtures/shared/list", options.ListObjectCount)

	fixtures := FixtureManifest{
		SharedReadKeys:  sharedReadKeys,
		LocalReadKeys:   localReadKeys,
		OverwriteKeys:   overwriteKeys,
		DeleteKeys:      deleteKeys,
		ListPrefix:      listPrefix,
		ListObjectCount: options.ListObjectCount,
	}

	scenarios := []BucketScenario{
		{
			ID:              "normal",
			Label:           "normal",
			Kind:            "normal",
			Bucket:          MakeBucketName(options.Prefix, "normal"),
			SnapshotEnabled: false,
			ForkDepth:       0,
		},
		{
			ID:              "snapshot-root",
			Label:           "snapshot-root",
			Kind:            "snapshot-root",
			Bucket:          MakeBucketName(options.Prefix, "snapshot-root"),
			SnapshotEnabled: true,
			ForkDepth:       0,
		},
	}

	cleanupManifest := func() BenchmarkManifest {
		return BenchmarkManifest{
			RunID:       runID,
			CreatedAt:   NowRFC3339(),
			Options:     options,
			Fixtures:    fixtures,
			Scenarios:   scenarios,
			ArtifactDir: artifactDir,
		}
	}

	if err := clients.CreateBucket(ctx, scenarios[0].Bucket, false); err != nil {
		return BenchmarkManifest{}, err
	}
	if err := clients.CreateBucket(ctx, scenarios[1].Bucket, true); err != nil {
		_ = CleanupRun(ctx, clients, cleanupManifest())
		return BenchmarkManifest{}, err
	}

	for _, keySet := range [][]string{sharedReadKeys, overwriteKeys, deleteKeys, listKeys} {
		if err := seedKeys(ctx, clients, scenarios[0].Bucket, keySet, options.ObjectSizeBytes, "normal-seed"); err != nil {
			_ = CleanupRun(ctx, clients, cleanupManifest())
			return BenchmarkManifest{}, err
		}
		if err := seedKeys(ctx, clients, scenarios[1].Bucket, keySet, options.ObjectSizeBytes, "snapshot-root-seed"); err != nil {
			_ = CleanupRun(ctx, clients, cleanupManifest())
			return BenchmarkManifest{}, err
		}
	}

	parentScenario := scenarios[1]
	for depth := 1; depth <= options.MaxForkDepth; depth++ {
		snapshotVersion, err := clients.CreateSnapshot(ctx, parentScenario.Bucket, fmt.Sprintf("fork-depth-%d", depth))
		if err != nil {
			_ = CleanupRun(ctx, clients, cleanupManifest())
			return BenchmarkManifest{}, err
		}

		child := BucketScenario{
			ID:                    fmt.Sprintf("fork-depth-%d", depth),
			Label:                 fmt.Sprintf("fork-depth-%d", depth),
			Kind:                  "fork",
			Bucket:                MakeBucketName(options.Prefix, fmt.Sprintf("fork-%d", depth)),
			SnapshotEnabled:       true,
			ForkDepth:             depth,
			ParentScenarioID:      parentScenario.ID,
			ParentBucket:          parentScenario.Bucket,
			ParentSnapshotVersion: snapshotVersion,
		}

		snapshotEnabled, err := clients.CreateForkBucket(ctx, child.Bucket, parentScenario.Bucket, snapshotVersion)
		if err != nil {
			_ = CleanupRun(ctx, clients, cleanupManifest())
			return BenchmarkManifest{}, err
		}
		child.SnapshotEnabled = snapshotEnabled
		if !snapshotEnabled && depth < options.MaxForkDepth {
			_ = CleanupRun(ctx, clients, cleanupManifest())
			return BenchmarkManifest{}, fmt.Errorf("fork bucket %s could only be created without snapshot support; depth %d+ requires snapshot-enabled forks", child.Bucket, depth+1)
		}

		scenarios = append(scenarios, child)
		parentScenario = child
	}

	for _, scenario := range scenarios {
		localScenarioKeys := make([]string, 0, len(localReadKeys))
		for _, key := range localReadKeys {
			localScenarioKeys = append(localScenarioKeys, scenarioLocalKey(key, scenario.ID))
		}
		if err := seedKeys(ctx, clients, scenario.Bucket, localScenarioKeys, options.ObjectSizeBytes, "local-seed:"+scenario.ID); err != nil {
			_ = CleanupRun(ctx, clients, cleanupManifest())
			return BenchmarkManifest{}, err
		}
	}

	manifest := BenchmarkManifest{
		RunID:       runID,
		CreatedAt:   NowRFC3339(),
		Options:     options,
		Fixtures:    fixtures,
		Scenarios:   scenarios,
		ArtifactDir: artifactDir,
	}
	if err := WriteJSON(fmt.Sprintf("%s/manifest.json", artifactDir), manifest); err != nil {
		_ = CleanupRun(ctx, clients, manifest)
		return BenchmarkManifest{}, err
	}
	return manifest, nil
}

func CleanupRun(ctx context.Context, clients *Clients, manifest BenchmarkManifest) error {
	scenarios := append([]BucketScenario(nil), manifest.Scenarios...)
	for i := 0; i < len(scenarios); i++ {
		for j := i + 1; j < len(scenarios); j++ {
			if scenarios[j].ForkDepth > scenarios[i].ForkDepth {
				scenarios[i], scenarios[j] = scenarios[j], scenarios[i]
			}
		}
	}
	for _, scenario := range scenarios {
		if err := clients.RemoveBucket(ctx, scenario.Bucket); err != nil {
			return err
		}
	}
	return nil
}

File: src/provision.ts
import { createPayload, ensureDir, makeBucketName, makeRunId, resolveArtifactDir, writeJson } from './util.js';
import { createBucketOrThrow, createForkBucketOrThrow, createSnapshotOrThrow, putObjectOrThrow, removeBucketOrThrow } from './tigris.js';
import type { AuthConfig, BenchmarkManifest, BenchmarkOptions, BucketScenario, FixtureManifest } from './types.js';

function rangeKeys(prefix: string, count: number): string[] {
  return Array.from({ length: count }, (_, index) => `${prefix}/obj-${String(index).padStart(4, '0')}.bin`);
}

async function seedKeys(bucket: string, keys: string[], options: { auth: AuthConfig; objectSizeBytes: number; seedPrefix: string }): Promise<void> {
  for (const key of keys) {
    const body = createPayload(options.objectSizeBytes, `${options.seedPrefix}:${bucket}:${key}`);
    await putObjectOrThrow(bucket, key, body, options.auth, true);
  }
}

export async function provisionRun(auth: AuthConfig, options: BenchmarkOptions): Promise<BenchmarkManifest> {
  const runId = makeRunId(options.prefix);
  const artifactDir = resolveArtifactDir(options.artifactsRoot, runId);
  await ensureDir(artifactDir);
  const mutableFixtureCount = options.iterations + options.warmup;

  const sharedReadKeys = rangeKeys('fixtures/shared/read', Math.max(options.iterations, options.sharedReadCount));
  const localReadKeys = rangeKeys('fixtures/local/read', Math.max(options.iterations, options.localReadCount));
  const overwriteKeys = rangeKeys('fixtures/shared/overwrite', Math.max(mutableFixtureCount, options.overwriteCount));
  const deleteKeys = rangeKeys('fixtures/shared/delete', Math.max(mutableFixtureCount, options.deleteCount));
  const listPrefix = 'fixtures/shared/list/';
  const listKeys = rangeKeys(listPrefix.replace(/\/$/, ''), options.listObjectCount);

  const fixtures: FixtureManifest = {
    sharedReadKeys,
    localReadKeys,
    overwriteKeys,
    deleteKeys,
    listPrefix,
    listObjectCount: options.listObjectCount,
  };

  const scenarios: BucketScenario[] = [];
  const normalScenario: BucketScenario = {
    id: 'normal',
    label: 'normal',
    kind: 'normal',
    bucket: makeBucketName(options.prefix, 'normal'),
    snapshotEnabled: false,
    forkDepth: 0,
  };
  scenarios.push(normalScenario);

  const snapshotRootScenario: BucketScenario = {
    id: 'snapshot-root',
    label: 'snapshot-root',
    kind: 'snapshot-root',
    bucket: makeBucketName(options.prefix, 'snapshot-root'),
    snapshotEnabled: true,
    forkDepth: 0,
  };
  scenarios.push(snapshotRootScenario);

  try {
    await createBucketOrThrow(normalScenario.bucket, auth, { enableSnapshot: false });
    await createBucketOrThrow(snapshotRootScenario.bucket, auth, { enableSnapshot: true });

    for (const keySet of [sharedReadKeys, overwriteKeys, deleteKeys, listKeys]) {
      await seedKeys(normalScenario.bucket, keySet, {
        auth,
        objectSizeBytes: options.objectSizeBytes,
        seedPrefix: 'normal-seed',
      });
      await seedKeys(snapshotRootScenario.bucket, keySet, {
        auth,
        objectSizeBytes: options.objectSizeBytes,
        seedPrefix: 'snapshot-root-seed',
      });
    }

    let parentScenario = snapshotRootScenario;
    for (let depth = 1; depth <= options.maxForkDepth; depth += 1) {
      const snapshotVersion = await createSnapshotOrThrow(parentScenario.bucket, auth, `fork-depth-${depth}`);
      const childScenario: BucketScenario = {
        id: `fork-depth-${depth}`,
        label: `fork-depth-${depth}`,
        kind: 'fork',
        bucket: makeBucketName(options.prefix, `fork-${depth}`),
        snapshotEnabled: true,
        forkDepth: depth,
        parentScenarioId: parentScenario.id,
        parentBucket: parentScenario.bucket,
        parentSnapshotVersion: snapshotVersion,
      };

      const forkSnapshotEnabled = await createForkBucketOrThrow(childScenario.bucket, parentScenario.bucket, snapshotVersion, auth);
      childScenario.snapshotEnabled = forkSnapshotEnabled;
      if (!forkSnapshotEnabled && depth < options.maxForkDepth) {
        throw new Error(
          `Fork bucket ${childScenario.bucket} could only be created without snapshot support. Depth ${depth + 1}+ requires snapshot-enabled fork buckets, so this run cannot continue.`,
        );
      }
      scenarios.push(childScenario);
      parentScenario = childScenario;
    }

    for (const scenario of scenarios) {
      const scenarioLocalKeys = localReadKeys.map((key) => key.replace('/read/', `/read/${scenario.id}/`));
      await seedKeys(scenario.bucket, scenarioLocalKeys, {
        auth,
        objectSizeBytes: options.objectSizeBytes,
        seedPrefix: `local-seed:${scenario.id}`,
      });
    }

    const manifest: BenchmarkManifest = {
      runId,
      createdAt: new Date().toISOString(),
      options,
      fixtures,
      scenarios,
      artifactDir,
    };

    await writeJson(`${artifactDir}/manifest.json`, manifest);
    return manifest;
  } catch (error) {
    await cleanupRun(auth, {
      runId,
      createdAt: new Date().toISOString(),
      options,
      fixtures,
      scenarios,
      artifactDir,
    });
    throw error;
  }
}

export async function cleanupRun(auth: AuthConfig, manifest: BenchmarkManifest): Promise<void> {
  const scenarios = [...manifest.scenarios].sort((left, right) => right.forkDepth - left.forkDepth);

  for (const scenario of scenarios) {
    try {
      await removeBucketOrThrow(scenario.bucket, auth);
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      console.warn(`Cleanup warning for ${scenario.bucket}: ${message}`);
    }
  }
}
